### ISSUE TITLE
A more ergonomic enumeration representation

### ISSUE THREAD

**Issue by sfackler** [danburkert/prost#276]
To deal with the "openness" of Protobuf enumeration types, Prost currently represents them in the message struct as `i32` and generates a Rust enum that can convert to and from `i32`. This is a bit awkward to work with, though.

A better approach could be to use a newtype with associated constants. To use the [`Syntax` enum](https://docs.rs/prost-types/0.6.1/prost_types/enum.Syntax.html) as an example, it would look like this:

```rust
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ...)]
pub struct Syntax(i32);

impl Syntax {
    pub const PROTO2: Syntax = Syntax(0);
    pub const PROTO3: Syntax = Syntax(1);

    // alternatively the field could just be public
    pub fn from_raw(raw: i32) -> Syntax {
        Syntax(raw)
    }

    pub fn as_raw(self) -> i32 {
        self.0
    }
}
```

This approach has a couple of advantages:

* It can be embedded directly in the message struct.
* It's more type safe since you normally won't have to interact with the raw `i32` value.
* You can still use it roughly like a Rust enum, and do things like match on the variants:
```rust
match syntax {
    Syntax::PROTO2 => { ... }
    Syntax::PROTO3 => { ... }
    syntax => panic!("unknown syntax variant {}", syntax.as_raw()),
}
```

**Comment by danburkert** [comment]
@sfackler are you aware that type-safe getters are generated for enum fields?  For instance: https://docs.rs/prost-types/0.6.1/prost_types/struct.Type.html#method.syntax

**Comment by danburkert** [comment]
I do like this idea, though.  It definitely has some merit.  My experience has been that working with enums through the typesafe getter/setter is 'good enough', but they have a discoverability problem.  I don't think they are actually documented anywhere other than the resulting rustdoc.

**Comment by sfackler** [comment]
Ooh, I did not know about those!

**Comment by sfackler** [comment]
I'd be happy to build out the implementation, but I'd be interested on your thoughts on how to structure it. 0.6 is pretty new, so this could just be a breaking change and prost bumps to 0.7, or it could end up as a new flag in prost_build::Config to allow users to opt into the new approach.

**Comment by jakeswenson** [comment]
@sfackler @danburkert How's this going? 
I have been using prost more and more, and I do find the enums _slightly_ painful; I think this proposal goes a long way to easing that.

**Comment by sfackler** [comment]
I'm going to work on this over the weekend.

**Comment by sfackler** [comment]
For an enum:

```proto
enum SomeEnum {
    SOME_ENUM_FOO = 0;
    SOME_ENUM_BAR = 2;
}
```

prost-build will produce:

```rust
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct SomeEnum(i32);

#[prost::enumeration]
impl SomeEnum {
    pub const FOO: SomeEnum = SomeEnum(0);
    pub const BAR: Self = SomeEnum(2);
}
```

and prost-derive will expand that to:

```rust
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct SomeEnum(i32);

impl SomeEnum {
    pub const FOO: SomeEnum = SomeEnum(0);
    pub const BAR: Self = SomeEnum(2);

    pub fn from_i32(value: i32) -> Option<SomeEnum> {
        match value {
            0 => Some(SomeEnum::FOO),
            2 => Some(SomeEnum::BAR),
            _ => None,
    }
}

impl Default for SomeEnum {
    fn default() -> SomeEnum {
        SomeEnum::FOO
    }
}

impl From<SomeEnum> for i32 {
    fn from(value: SomeEnum) -> i32 {
        value.0
    }
}
```

**Comment by nw0** [comment]
Hi @sfackler, I've run into the same issue but found this before writing my own code. Is this currently WIP? I'd be happy to help out if so.

**Comment by sfackler** [comment]
I spent a while working on it, but then got sidetracked with other things I've pushed what I have onto a branch but I'm not sure when I'll be able to circle back to it: https://github.com/sfackler/prost/commit/0330d92fcb81761c97780326bdc3f43a5649efd5

**Comment by danburkert** [comment]
Can anyone share some application code that this alternate representation cleans up?  I'd love to be convinced since this seems to come up over and over, but I'm still not seeing how this is materially better than the generated getters + generated (Rust) enums.

**Comment by mieubrisse** [comment]
Are the generated getters still working? I have this Prost-generated code:

```rust
#[derive(Clone, PartialEq, ::prost::Message)]
pub struct SuiteRegistrationResponse {
    #[prost(enumeration = "SuiteAction", tag = "1")]
    pub suite_action: i32,
}
/// Tells the suite what action it should perform, based on the args that the API container received
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
#[repr(i32)]
pub enum SuiteAction {
    /// Indicates that the testsuite should operate in metadata-serializing mode, sending suite metadata to the
    ///  API container
    SerializeSuiteMetadata = 0,
    /// Indicates that the testsuite should operate in test-executing mode, running a test
    ExecuteTest = 1,
}
```

which is throwing the following where when I try to use `from`:

<img width="777" alt="Screen Shot 2021-02-10 at 12 12 38 PM" src="https://user-images.githubusercontent.com/1142185/107552570-70750500-6bb2-11eb-8eba-1b3e0198773e.png">

**Comment by danburkert** [comment]
Yes, they are still present in `prost` 0.7.  `prost` enums have never provided a `From<i32>` impl, I think you may be looking for the `from_i32` method.  Or you can use the `SuiteRegistrationResponse::suite_action()` getter.

**Comment by mieubrisse** [comment]
Gotcha, `from_i32` was the missing bit (weirdly VS Code wasn't showing that as an option). Using that method works now!

**Comment by danburkert** [comment]
Yeah these methods are generated via derives, so RA doesn't seem to 'see' through it a lot of the time.  Best way to work with `prost` in my experience is to generate the docs and look there; rust-doc will show the derive generated code.

**Comment by avsaase** [comment]
Maybe I'm missing something here but why not just generate a Rust enum with the `#[non_exhaustive]` attribute?

**Comment by benma** [comment]
The discoverability problem is pretty big imo. I want to expose a Prost generated struct as a function param in a library function. The rust docs just says `i32` for the enum type. It would be much preferable if it actually showed the enum type and all variants. 

@danburkert could we have a decision to move forward? There are multiple proposals better than the status quo, e.g [Unknown(i32)](https://github.com/tokio-rs/prost/issues/303#issue-602485729) or the proposal at the top of this issue.

**Comment by mzabaluev** [comment]
@avsaase
> Maybe I'm missing something here but why not just generate a Rust enum with the `#[non_exhaustive]` attribute?

In proto3, the enums are [open](https://protobuf.dev/programming-guides/proto3/#enum) and the language guide suggests that unknown values are preserved in the message. I could not find anything more normative, but Prost currently conforms to this by erasing the enum type in message fields. While I mostly find myself needing more idiomatic domain types to work with protobuf-originated data anyway, this gives more room for programmer error than necessary.

I lean towards this proposal over an added `Unknown(i32)` variant, since that one:

1. injects a magical name which might come into conflict with an actually defined `*UNKNOWN` variant;
2. loses `#[repr(i32)]` and makes the size of all generated enum types to be two 32-bit words rather than one.

A somewhat more cumbersome solution that preserves the enums could be a generic wrapper like this:

```rust
pub enum EnumFieldValue<T> {
    Known(T),
    Unknown(i32),
}

impl<T> EnumFieldValue<T> {
    pub fn unwrap(self) -> T {
        unimplemented!("for brevity")
    }

    pub fn known_or_else<E, F>(self, err: F) -> Result<T, E>
    where
        F: FnOnce(i32) -> E,
    {
        unimplemented!("for brevity")
    }
}
```

This has the same effect on the field size as `Unknown(i32)`.

Regardless of the above, you'd want your proto3-derived enums to be annotated with `#[non_exhaustive]`, as you'd want your [structs](https://github.com/tokio-rs/prost/issues/399#issuecomment-737436729) (with some builder code to get over the initialization issues) since this is how protobuf bindings are expected to be evolved with their semver conventions.

**Comment by mzabaluev** [comment]
> This has the same effect on the field size as `Unknown(i32)`.

`PhantomData` to the rescue: 

```rust
pub struct EnumFieldValue<T> {
    value: i32,
    _phantom: core::marker::PhantomData<T>,
}

impl<T> EnumFieldValue<T>
where
    i32: TryInto<T>,
{
    pub fn unwrap(self) -> T {
        let v = self.value;
        v.try_into()
            .map_err(|_| format!("unknown field value {}", v))
            .unwrap()
    }

    pub fn known_or_else<E, F>(self, err: F) -> Result<T, E>
    where
        F: FnOnce(i32) -> E,
    {
        let v = self.value;
        v.try_into().map_err(|_| err(v))
    }
}
```

**Comment by ryankurte** [comment]
every couple of years i pick prost back up and run into this again 😅

> The discoverability problem is pretty big imo. I want to expose a Prost generated struct as a function param in a library function. The rust docs just says i32 for the enum type. It would be much preferable if it actually showed the enum type and all variants.

real enum types are definitely better for documentation / external use / compatibility with things like serde!

given fields are wrapped in `Option`s already would it be unreasonable for prost_build to provide a lossy mode where empty _and_ unknown enum variants become `None`? that way at least folks who don't need access to unknown enum variants could opt-in to having well formed rust types.

**Comment by caspermeijn** [comment]
> real enum types are definitely better for documentation / external use / compatibility with things like serde!

I agree, we should work towards using the actual enum for enum fields.
 
> given fields are wrapped in `Option`s already would it be unreasonable for prost_build to provide a lossy mode for enums where empty _and_ unknown variants get promoted to `None`? that way folks who don't need access to unknown enum variants could opt-in to having well formed rust types.

I would argue that separated lossy mode is also not discoverable. I think `prost-build` should generate better default code. 

I think a good middle ground for all described use cases is: 

```rust
pub enum EnumOption<T> {
    Known(T),
    Unknown(i32),
    Missing,
}

impl<T> EnumFieldValue<T> {
    pub fn unwrap(self) -> T {
        todo!()
    }

    pub fn known(self) -> Option<T> {
        todo!()
    }
}
```
And then let `prost-build` generate the following code:

```rust
#[repr(i32)]
pub enum Label {
        Optional = 1,
        Required = 2,
        Repeated = 3,
}

#[derive(::prost::Message)]
pub struct FieldDescriptorProto {
    #[prost(int32, optional, tag = "3")]
    pub label: ::prost::EnumOption<Label>,
}
```

@ryankurte Are you willing to do the work? I suggest opening a proof of concept pull request with the `enum EnumOption` and a work in progress code generator. Then we can review that work before you spend all the work refactoring the test suite. Feel free to choose better names.

**Comment by mzabaluev** [comment]
I have just submitted #1061 which I have been holding off for some time. I will un-draft it after adding documentation. The memory layout of the `OpenEnum` wrapper is exactly that of `i32`, so this does not add overhead unlike the enum-based solutions.

One thing I dislike about the `EnumOption` proposal above is that it discards the idiomatic `Option` type, which is used for other non-scalar fields. Memory layout optimization is a wrong reason to do it, because composing `Option` with an inner `Known` | `Unknown` enum will use niche value optimization, so you only need to pay the overhead of a single enum discriminant if you decide to encode these three cases with Rust-native enums.

Furthermore, I'm not convinced that making all enum fields unconditionally optional is the best ergonomic approach in all cases, for the same reason why scalars are not guarded with an `Option` by default. There is a [protobuf style guideline](https://protobuf.dev/programming-guides/style/#enums) backed by a `buf` lint, where the zero-valued variant is explicitly designated as "unspecified" and used as a fallback for the missing field in e.g. generated Go code.

**Comment by ryankurte** [comment]
> I would argue that separated lossy mode is also not discoverable. I think prost-build should generate better default code.

definitely in favour of more useful / idiomatic generated code! it'd be nice to be able to use prost-generated objects like they're normal(ish) rust objects. something like `EnumOption` seems reasonable, the only caveat being that folks may need to be able to control the way this serialises (though i think it would be reasonable to set it to be transparent as a baseline and worry about customising this later).

@mzabaluev the `OpenEnum` trick is clever but doesn't improve the representation or use of enums with match / in other objects. also this makes doing things like `ToString`/`FromString`/[`VariantNames`](https://docs.rs/strum/latest/strum/derive.VariantNames.html) useful for serialisation and command-line stuff even weirder than it is now (though arguably this could be paved over with generated code).

> Furthermore, I'm not convinced that making all enum fields unconditionally optional is the best ergonomic approach in all cases, ...

i think your earlier proposal would make sense with separate handling of known/unknown and [optional](https://protobuf.dev/programming-guides/proto3/#field-labels) fields? so `Option` is only used where required to express [field presence](https://protobuf.dev/programming-guides/field_presence/#presence-in-proto3-apis), otherwise you just get the normal wrapper.

```rust
pub enum EnumFieldValue<T: Default> {
  Known(T),
  Unknown(i32),
}

struct Something {
  // Normal fields are always optional and default to 0'th variant
  pub normal_field: EnumFieldValue<SomeEnum>,
  // Field labeled 'optional' should be able to represent zero _or_ unset condition
  pub optional_field: Option<EnumFieldValue<AnotherEnum>>,
}
```

**Comment by mzabaluev** [comment]
> the `OpenEnum` trick is clever but doesn't improve the representation or use of enums with match / in other objects.

I think you can match on `message.field.known()` in most cases, albeit less conveniently (matching on specific unknown values should be considered pathological). The `OpenEnum` wrapper is not meant to be used in objects other than the generated message structs, you should use the enum type itself for that. It can be cumbersome when the whole message struct value is matched as part of a composite type, `Result` being perhaps the most usual case of this.

> also this makes doing things like `ToString`/`FromString`/[`VariantNames`](https://docs.rs/strum/latest/strum/derive.VariantNames.html) useful for serialisation and command-line stuff even weirder than it is now (though arguably this could be paved over with generated code).

Regardless of the wrapper's design, such implementations/derives would need to be transparent for the known values. I don't think you can do `Display`/`ToString` infallibly on open enum values, as there is no universally accepted rendering for the unknown cases.

**Comment by mzabaluev** [comment]
> i think your earlier proposal would make sense with separate handling of known/unknown and [optional](https://protobuf.dev/programming-guides/proto3/#field-labels) fields? so `Option` is only used where required to express [field presence](https://protobuf.dev/programming-guides/field_presence/#presence-in-proto3-apis) for encode and decode, otherwise you just get the `EnumFieldValue` wrapper with defaults / elision.

If nobody's working on it, I can try to remake my `OpenEnum` wrapper as the proper enum over the weekend.
This effectively computes the known/unknown discriminant once on construction and enables destructuring of enum fields as part of a message. The memory overhead should be tolerable outside, perhaps, of very memory-constrained embedded uses.

**Comment by ArjixWasTaken** [comment]
Since proto is pretty open on enum values, meaning they must be preserved even when invalid,
why not wrap them in a Result?

I think `Result<T, i32>` where `T` is the generated enum would be more than enough

**Comment by mzabaluev** [comment]
Slice matching is one case that makes `Vec<OpenEnum<T>>` more ergonomical than an opaque representation.
